_source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md

```diff
@@ -0,0 +1,362 @@
+
+To start your application, run the Uvicorn server:
+
+```bash
+uvicorn main:app --reload
+```
+
+Then, open [http://127.0.0.1:8000](http://127.0.0.1:8000) in your browser to see your "Hello, world" endpoint in action:
+
+![A "Hello, World" endpoint in FastAPI](https://i.imgur.com/XHwSIuy.png)
+
+FastAPI's interactive documentation is available at [http://127.0.0.1:8000/docs](http://127.0.0.1:8000/docs). Once you're there, you'll see details about each endpoint automatically. FastAPI uses type hints and context in your application to produce these docs on its own.
+
+![FastAPI interactive documentation](https://i.imgur.com/Va63heZ.png)
+
+Now that you've seen how easy it is to get started, you're ready to build a more useful application. In the remainder of this tutorial, you'll see how to create protected endpoints in FastAPI that use Okta as an authorization server.
+
+## Securing FastAPI with Okta
+
+To demonstrate some of the more advanced features of the FastAPI framework, I'll show you how to build a protected endpoint that uses [the client credentials flow](https://developer.okta.com/docs/guides/implement-client-creds/overview/) to authorize access. By the end of this tutorial, you will be able to enter your Okta client ID and secret into FastAPI's interactive docs to get an access token. You'll see how to validate this JWT access token remotely using Okta's [introspect](https://developer.okta.com/docs/reference/api/oidc/#introspect) endpoint and locally using the [Okta JWT Python library](https://github.com/adithyasampatoor/okta_jwt).
+
+If you'd like to run the final application, [the code is available on GitHub](https://github.com/oktadeveloper/okta-fastapi), or you can follow along for step-by-step instructions.
+
+### Setting Up a New Application in Okta
+
+Before you create any endpoints in your FastAPI application, you'll need to create a new application in Okta and get your Authorization Server's issuer URL and audience. As you go through these steps, add the Okta environment variables to a new file in your application called `.env`. You'll see how to use these variables later in the tutorial.
+
+To create a new server application, log in to your Okta account and go to **Applications** and click the **Add Application** button in the top left. Select **Service, Machine-to-Machine**, then click **Next**.
+
+![Creating a new service application in Okta](https://i.imgur.com/8EUhPMx.png)
+
+Enter a name for your application and click **Next** again.
+
+![Entering a name for your application in Okta](https://i.imgur.com/IS1Z4jn.png)
+
+Copy the **Client ID** and **Client Secret** from this page and add them to your FastAPI application's `.env` file as `OKTA_CLIENT_ID` and `OKTA_CLIENT_SECRET` respectively.
+
+Your `.env` file should look like the example below, with your `OKTA_CLIENT_ID` and `OKTA_CLIENT_SECRET` values filled out:
+
+```
+OKTA_CLIENT_ID=
+OKTA_CLIENT_SECRET=
```

Comment:
We typically do placeholders like `{yourClientID}`, but it's probably fine to leave like this.